accounts/accounts/routes/ui.py

```diff
@@ -0,0 +1,155 @@
+
+blueprint = Blueprint('ui', __name__, url_prefix='/user')
+
+
+def user_is_owner(session: domain.Session, user_id: str, **kw: Any) -> bool:
+    """Determine whether the authenticated user matches the requested user."""
+    return bool(session.user.user_id == user_id)
+
+
+@blueprint.before_request
+def get_next_page() -> None:
+    """Get the next page; this is where we'll redirect the user on success."""
+    request.next_page = request.args.get('next_page', '')
+
+
+def anonymous_only(func: Callable) -> Callable:
+    """Redirect logged-in users to their profile."""
+    @wraps(func)
+    def wrapper(*args: Any, **kwargs: Any) -> Any:
+        if request.session:
+            user = request.session.user
+            target = url_for('ui.view_profile', user_id=user.user_id)
+            content = redirect(target, code=status.HTTP_303_SEE_OTHER)
+            response = make_response(content)
+            return response
+        return func(*args, **kwargs)
+    return wrapper
+
+
+def set_cookies(response: Response, data: dict) -> None:
+    """
+    Update a :class:`.Response` with cookies in controller data.
+
+    Contollers seeking to update cookies must include a 'cookies' key
+    in their response data.
+    """
+    # Set the session cookie.
+    cookies = data.pop('cookies')
+    if cookies is None:
+        return None
+    for cookie_key, (cookie_value, expires) in cookies.items():
+        cookie_name = current_app.config[f'{cookie_key.upper()}_NAME']
+        response.set_cookie(cookie_name, cookie_value, expires=expires)
+
+
+@blueprint.route('/register', methods=['GET', 'POST'])
+@anonymous_only
+def register() -> Response:
+    """Interface for creating new accounts."""
+    captcha_secret = current_app.config['CAPTCHA_SECRET']
+    ip_address = request.remote_addr
+    data, code, headers = registration.register(request.method, request.form,
+                                                captcha_secret, ip_address)
+
+    # Flask puts cookie-setting methods on the response, so we do that here
+    # instead of in the controller.
+    if code is status.HTTP_201_CREATED:
+        target = url_for('ui.view_profile', user_id=data['user_id'])
+        response = make_response(redirect(target, code=code))
+        set_cookies(response, data)
+        return response
+    content = render_template("accounts/register.html", **data)
+    response = make_response(content, code, headers)
+    return response
+
+
+@blueprint.route('/<string:user_id>/profile', methods=['GET'])
+@scoped(scopes.VIEW_PROFILE, authorizer=user_is_owner)
+def view_profile(user_id: str) -> Response:
+    """User can view their account information."""
+    data, code, headers = registration.view_profile(user_id, request.session)
+    return render_template("accounts/profile.html", **data)
+
+
+@blueprint.route('/<string:user_id>/profile/edit', methods=['GET', 'POST'])
+@scoped(scopes.EDIT_PROFILE, authorizer=user_is_owner)
```

Comment:
Here's an example of the ``scoped`` decorator in action. The ``EDIT_PROFILE`` scope is required to access this route, and in addition the authenticated user may only edit their own profile.